### DISCUSSION TITLE
AList.find /AList.tryFind, etc ?

### DISCUSSION THREAD

**Discussion by somelinguist** [fsprojects/FSharp.Data.Adaptive#110] [Q&A]
Hi, I'm still getting familiar with the library. I was wondering if there is a reason `AList.find`, `AList.tryFind`, `AList.findIndex`, `AList.tryFindIndex`, etc. weren't part of the library? Maybe also `pick`/`tryPick`? 

I see that IndexList has implementations for them (except plain `pick`), but was wondering if there were performance reasons for not having them for AList. 

Is there some other way I should use to make an adaptive value from a single item matching a predicate from a clist/alist?

The closest I've come up with so far is `AList.filter |> AList.tryFirst`, but I'm assuming that goes through the whole list?

Thanks!

**Comment by krauthaufen** [comment]
Hey,

The things you want are operations on the entire list (at least conceptually) and we simple never had the need for them.
I think they could be implemented "incrementally" s.t. changes won't need to re-evaluate everything. 

I'll look into that once i have a little time

**Reply by somelinguist** [reply]
Thanks for the info. For context, what I'm trying to do have an adaptive value for the currently selected nested item. Something like `Sentences[sid].Parts[pid]` where `sid` and `pid` are themselves adaptive values the represent the index for the currently selected item. `sid` and `pid` are really more like keys the way I've implemented them, but they're ordered, which I is why was still looking at using clist/alist instead of cmap/amap. Since they're unique, I'm going through each list, and once I find sid or pid, I know I have the only one.

**Reply by krauthaufen** [reply]
We use our `Index` for exactly that, it's comparable and has proper equality and serves as key for alist/clist/etc

**Reply by somelinguist** [reply]
Are`Index`s generic/extendable so that I could I could use my own data type that implements `IComparable`?

**Reply by krauthaufen** [reply]
Hey, currently not, but we could theoretically create something like `amap` using sorted maps instead of hashmaps. I dont thinkt this would be too hard, because all the amap implementations would basically work for that too.
Would that help in your case?

**Reply by somelinguist** [reply]
Hey. Thanks for the quick reply. That would probably work great. I just need something that can be traversed in a predictable/sorted order and also be used for lookups. I’ve usually just used the `find` kind of functions  on an appropriate ordered collection with some kind of id/key for the lookups, but a sorted/ordered map also sounds very helpful.
